src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -4,58 +4,203 @@
+            CodeActionParams request,
             Document document,
             ICodeFixService codeFixService,
             ICodeRefactoringService codeRefactoringService,
+            IThreadingContext threadingContext,
             LSP.Range selection,
             CancellationToken cancellationToken)
         {
-            var (codeFixCollections, codeRefactorings) = await GetCodeFixesAndRefactoringsAsync(
-                document, codeFixService,
-                codeRefactoringService, selection,
-                cancellationToken).ConfigureAwait(false);
+            var actionSets = await GetActionSetsAsync(
+                document, codeFixService, codeRefactoringService, threadingContext, selection, cancellationToken).ConfigureAwait(false);
+            if (!actionSets.HasValue)
+            {
+                return Array.Empty<VSCodeAction>();
+            }
+
+            var _ = ArrayBuilder<VSCodeAction>.GetInstance(out var codeActions);
+            foreach (var set in actionSets)
+            {
+                foreach (var suggestedAction in set.Actions)
+                {
+                    // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+                    if (suggestedAction.CodeAction is CodeActionWithOptions)
+                    {
+                        continue;
+                    }
+
+                    codeActions.Add(GenerateVSCodeAction(
+                        request, GetNestedActionsFromActionSet(suggestedAction),
+                        GetCodeActionKindFromSuggestedActionCategoryName(set.CategoryName!)));
+                }
+            }
+
+            return codeActions.ToArray();
+        }
+
+        private static VSCodeAction GenerateVSCodeAction(
+            CodeActionParams request,
+            CodeAction codeAction,
+            CodeActionKind codeActionKind,
+            string currentTitle = "")
+        {
+            using var _ = ArrayBuilder<VSCodeAction>.GetInstance(out var nestedActions);
+
+            if (!string.IsNullOrEmpty(currentTitle))
+            {
+                // Adding a delimiter for nested code actions, e.g. 'Suppress or Configure issues|Suppress IDEXXXX|in Source'
+                currentTitle += '|';
+            }
+
+            currentTitle += codeAction.Title;
+
+            // Nested code actions' unique identifiers consist of: parent code action unique identifier + '|' + title of code action
+            foreach (var action in codeAction.NestedCodeActions)
+            {
+                nestedActions.Add(GenerateVSCodeAction(request, action, codeActionKind, currentTitle));
+            }
+
+            // WPF uses the first underscore to create keyboard shortcuts, so we have to insert an extra underscore
+            // wherever there's an underscore for it to be displayed properly.
+            var title = codeAction.Title.Replace("_", "__");
```

Comment:
does this not need to be done for the code actions we provide to VS?

Or alternatively, it seems like the LSP client should handle this.  Right now the client displaying code actions is WPF based, but that may not always be true.